nemo/collections/common/data/dataset.py

```diff
@@ -286,3 +289,313 @@ def __len__(self):
+            # need to use numpy instead of torch here because we need numpy's trim_zeros function
+            created_sample_audios.append(audio.cpu().numpy())
+            comp_text = torch.cat([comp_text, labels], dim=0)
+
+            if pure_mono:
+                break
+
+        sample_channels = list(set([s.ndim for s in created_sample_audios]))
+        if len(sample_channels) > 1:
+            raise RuntimeError(
+                "Mixture of audios with different number of channels in CodeSwitchedDataset. All sources must be same number of channels."
+            )
+
+        multichannel = sample_channels[0] > 1
+
+        if multichannel:
+            comp_audio = np.zeros(
+                shape=(int(self.pause_start * self.sample_rate / 1000.0), created_sample_audios[0].shape[-1]),
+                dtype=created_sample_audios[0].dtype,
+            )
+        else:
+            comp_audio = np.zeros(
+                shape=(int(self.pause_start * self.sample_rate / 1000.0),), dtype=created_sample_audios[0].dtype
+            )
+
+        for idx, wav in enumerate(created_sample_audios):
+            if not multichannel:
+                wav = np.trim_zeros(wav)
+
+            wav_norm = wav * (10.0 ** (self.db_norm / 20.0) / np.maximum(0.01, (wav ** 2).mean(axis=0) ** 0.5))
+
+            if idx < len(created_sample_audios) - 1:
+                if multichannel:
+                    wav_norm = np.append(
+                        wav_norm,
+                        np.zeros(
+                            shape=(
+                                int(self.pause_join * self.sample_rate / 1000.0),
+                                created_sample_audios[0].shape[-1],
+                            ),
+                            dtype=comp_audio.dtype,
+                        ),
+                        axis=0,
+                    )
+                else:
+                    wav_norm = np.append(
+                        wav_norm,
+                        np.zeros(shape=(int(self.pause_join * self.sample_rate / 1000.0),), dtype=comp_audio.dtype),
+                        axis=0,
+                    )
+
+            comp_audio = np.append(comp_audio, wav_norm, axis=0)
+
+        if multichannel:
+            comp_audio = np.append(
+                comp_audio,
+                np.zeros(
+                    shape=(int(self.pause_end * self.sample_rate / 1000.0), created_sample_audios[0].shape[-1]),
+                    dtype=comp_audio.dtype,
+                ),
+                axis=0,
+            )
+        else:
+            comp_audio = np.append(
+                comp_audio,
+                np.zeros(shape=(int(self.pause_end * self.sample_rate / 1000.0),), dtype=comp_audio.dtype),
+                axis=0,
+            )
+
+        if self.augmentor is not None:
+            # import here to avoid circular import error
+            from nemo.collections.asr.parts.preprocessing import AudioSegment
+
+            mb = io.BytesIO()
```

Comment:
This will be super slow right? Better to just not allow augmentation here then?

Reply:
So this is the part where we apply the augmentor to the final, synthetic utterance only, and why we always pass `Augmentor=None` to the individual language sub-datasets. We want all the individual samples from the individual mono-languages to be clean, and then we build the clean synthetic, and then we hit that final synthetic with the augmentation.

As for the logic, it's all in-memory operations with io.BytesIO, nothing is actually being written to disk. I have done several training runs with online CS and augmentation, and it doesn't slow it down by much at all, still very much usable.

There's sadly no way around it, we need to do augmentation on the final, full synthetic sample because many augmentations depend on features of the full waveform to work correctly (for example the new NormPerturbation).